atlasdb-backup/src/main/java/com/palantir/atlasdb/backup/cassandra/CqlCluster.java

```diff
@@ -0,0 +1,142 @@
+        InetSocketAddress firstHost = hosts.stream().findAny().orElseThrow();
+        Cluster cluster = new ClusterFactory(
+                        () -> Cluster.builder().withNettyOptions(new SocksProxyNettyOptions(firstHost)))
+                .constructCluster(hosts, config);
+        return new CqlCluster(cluster, config);
+    }
+
+    private static Set<InetSocketAddress> getHosts(CassandraKeyValueServiceConfig config) {
+        return config.servers().accept(new Visitor<>() {
+            @Override
+            public Set<InetSocketAddress> visit(DefaultConfig defaultConfig) {
+                // TODO(gs): fail instead here? can we assume CQL support?
+                return defaultConfig.thriftHosts();
+            }
+
+            @Override
+            public Set<InetSocketAddress> visit(CqlCapableConfig cqlCapableConfig) {
+                return cqlCapableConfig.cqlHosts();
+            }
+        });
+    }
+
+    public Map<InetSocketAddress, Set<TokenRange>> getTokenRanges(String tableName) {
+        try (Session session = cluster.connect()) {
+            Metadata metadata = session.getCluster().getMetadata();
+            String keyspaceName = config.getKeyspaceOrThrow();
+            KeyspaceMetadata keyspace = metadata.getKeyspace(keyspaceName);
+            TableMetadata tableMetadata = keyspace.getTable(tableName);
+            Set<Token> partitionTokens = getPartitionTokens(session, tableMetadata);
+            Map<InetSocketAddress, Set<TokenRange>> tokenRangesByNode =
+                    ClusterMetadataUtils.getTokenMapping(getHosts(config), metadata, keyspaceName, partitionTokens);
+
+            if (partitionTokens.isEmpty()) {
+                log.trace(
+                        "No token ranges identified requiring repair",
+                        SafeArg.of("keyspace", keyspace),
+                        SafeArg.of("table", tableName));
+            } else {
+                log.trace(
+                        "Identified token ranges requiring repair",
+                        SafeArg.of("keyspace", keyspace),
+                        SafeArg.of("table", tableName),
+                        SafeArg.of("numPartitionKeys", partitionTokens.size()),
+                        SafeArg.of("tokenRanges", tokenRangesByNode));
+
+                int numTokenRanges =
+                        tokenRangesByNode.values().stream().mapToInt(Set::size).sum();
+
+                log.debug(
+                        "Identified token ranges requiring repair",
+                        SafeArg.of("keyspace", keyspace),
+                        SafeArg.of("table", tableName),
+                        SafeArg.of("numPartitionKeys", partitionTokens.size()),
+                        SafeArg.of("numTokenRanges", numTokenRanges));
+            }
+
+            return tokenRangesByNode;
+        }
+    }
+
+    private static Set<Token> getPartitionTokens(Session session, TableMetadata tableMetadata) {
```

Comment:
utils copied over from internal backup product, but hopefully will not exist in that product once we're done.